core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -657,6 +688,43 @@ class BlockManagerMasterEndpoint(
     }
   }
 
+  private def getShufflePushMergerLocations(
+      numMergersNeeded: Int,
+      hostsToFilter: Set[String]): Seq[BlockManagerId] = {
+    val blockManagersWithExecutors = blockManagerIdByExecutor.groupBy(_._2.host)
+      .mapValues(_.head).values.map(_._2).toSet
+    val filteredBlockManagersWithExecutors = blockManagersWithExecutors
+      .filterNot(x => hostsToFilter.contains(x.host))
+    val filteredMergersWithExecutors = filteredBlockManagersWithExecutors.map(
+      x => BlockManagerId(x.executorId, x.host, StorageUtils.externalShuffleServicePort(conf)))
```

Comment:
Will the `executorId` part of shuffle push merger location ever be used?

(I do not think so, but if yes then be aware `mapValues(_.head)` on a HashMap won't produce a deterministic result.)

If not then the above lines can be simplified a lot, for example by introducing a constant ID for shuffle push merger locations. This might help in the debugging latter, too.

Reply:
Thanks for pointing this one out.
We currently use empty String for `executorId` part of a `BlockManagerId` for shuffle push merger locations.
This part should conform to the convention, and we should also do a dedup here since we could have multiple executors on the same host.